documents/MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,1081 @@
+
+## What’s new in V2 SDK
+
+* V2 SDK client is truly async. Operations take callback functions/lambdas, that is called when the operation is registered with the server
+    Blocking calls can be emulated by setting an `std::promise<>` in the callback by calling `promise.set_value() `and then waiting for the returned `std::future<>` object to be resolved after the function call by calling `future.wait()`
+* V2 SDK provides implementation for MQTT5 protocol, the next step in evolution of the MQTT protocol.
+* Public API terminology has changed. You `S``tart()` or `S``top()` the MQTT5 client rather than `Connect` or `Disconnect` like in V1. This removes the semantic confusion with the connect/disconnect as the client-level controls vs. internal recurrent networking events.
+* Support for Fleet Provisioning AWS IoT Core service.
+
+Public API for almost all actions and operations has changed significantly. For more details about the new features and to see specific code examples, refer to the other sections of this guide.
+
+
+## How To Get Started with V2 SDK
+
+There are differences between IoT C++ V1 SDK and IoT C++ V2 SDK. Below are changes you need to make to use IoT C++ V2 SDK features. For more information about MQTT5, visit [MQTT5 User Guide](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md).
+
+
+### MQTT Protocol
+
+V1 SDK uses an MQTT version 3.1.1 client under the hood.
+
+V2 SDK provides MQTT version 3.1.1 and MQTT version 5.0 client implementations. This guide focuses on the MQTT5 since this version is significant improvement over MQTT3. See MQTT5 features section.
+
+
+### Client Builder
+
+To access the AWS IoT service, you must initialize an MQTT client.
+
+In V1 SDK, the [awsiotsdk::MqttClient](http://aws-iot-device-sdk-cpp-docs.s3-website-us-east-1.amazonaws.com/latest/classawsiotsdk_1_1_mqtt_client.html) class represents an MQTT client. You instantiate the client directly passing all the required parameters to the class constructor. It’s possible to change client settings after its creation using `set*` methods, e.g. `SetAutoReconnectEnabled` or `SetMaxReconnectBackoffTimeout`.
+
+In V2 SDK, the [Aws::Iot::MqttClient](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_iot_1_1_mqtt_client.html) class represents an MQTT client, specifically MQTT5 protocol. V2 SDK provides an [Aws::Iot::Mqtt5ClientBuilder](https://aws.github.io/aws-iot-device-sdk-cpp-v2/class_aws_1_1_iot_1_1_mqtt5_client_builder.html) designed to easily create common configuration types such as direct MQTT or WebSocket connections.
+Once an MQTT5 client is built and finalized, the resulting MQTT5 client cannot have its settings modified.
+
+_Example of creating a client in V1_
+
+```
+#include "OpenSSLConnection.hpp"
```

Comment:
Code blocks should be marked as CPP code. It seems, all code blocks should be fixed.